addons/survey/controllers/survey_session_manage.py

```diff
@@ -9,21 +9,37 @@
 
 from odoo import fields, http
 from odoo.http import request
-from odoo.tools import is_html_empty
+from odoo.tools import consteq, is_html_empty
 
 
 class UserInputSession(http.Controller):
-    def _fetch_from_token(self, survey_token):
-        """ Check that given survey_token matches a survey 'access_token'.
+
+    def _fetch_from_token(self, survey_token=False, survey_id=False):
+        """ Check that given survey_token matches a survey 'access_token', or 'survey_id'.
         Unlike the regular survey controller, user trying to access the survey must have full access rights! """
-        return request.env['survey.survey'].search([('access_token', '=', survey_token)])
+        if survey_id:
+            survey = request.env['survey.survey'].browse(survey_id).exists()
+        else:
+            survey = request.env['survey.survey'].search([('access_token', '=', survey_token)])
 
-    def _fetch_from_session_code(self, session_code):
+        if survey and not consteq(survey.access_token, survey_token):
+            raise NotFound()
+        return survey
+
+    def _fetch_from_session_code(self, session_code=False, survey_id=False):
         """ Matches a survey against a passed session_code, and checks if it is valid.
         If it is valid, returns the start url. Else, the error type."""
-        if not session_code:
+        if not session_code and not survey_id:
             return None, {'error': 'survey_wrong'}
```

Comment:
I feel like you probably still want to fail if there's no session code though, as said above `token = False` could match anything, and if the session_code is False then it's probably not meant to be accessed